shedlock-provider-hazelcast/src/main/java/net/javacrumbs/shedlock/provider/hazelcast/HazelcastLockProvider.java

```diff
@@ -0,0 +1,191 @@
+ * Below, the mechanims :
+ * - The Lock, an instance of {@link HazelcastLock}, is obtained / created when :
+ * - the lock is not not already locked by other process (lock - referenced by its name - is not present in the Hazelcast locks store OR unlockable)
+ * - the lock is expired : {@link Instant#now()} > {@link HazelcastLock#unlockTime} where unlockTime have by default the same value of {@link HazelcastLock#lockAtMostUntil}
+ * and can have the value of {@link HazelcastLock#lockAtLeastUntil} if unlock action is used
+ * - expired object is removed
+ * - the lock is owned by not available member of Hazelcast cluster member
+ * - no owner objectis removed
+ * - Unlock action :
+ * - removes lock object when {@link HazelcastLock#lockAtLeastUntil} is not come
+ * - override value of {@link HazelcastLock#unlockTime} with {@link HazelcastLock#lockAtLeastUntil} (its default value is the same of {@link HazelcastLock#lockAtLeastUntil})
+ */
+public class HazelcastLockProvider implements LockProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(HazelcastLockProvider.class);
+
+    protected static final String LOCK_STORE_KEY_DEFAULT = "shedlock_storage";
+
+    /**
+     * Key used for get the lock container (an {@link IMap}) inside {@link #hazelcastInstance}.
+     * By default : {@link #LOCK_STORE_KEY_DEFAULT}
+     */
+    protected final String lockStoreKey;
+
+    /**
+     * Instance of the Hazelcast engine used by the application.
+     */
+    protected HazelcastInstance hazelcastInstance;
+
+    /**
+     * Instanciate the provider.
+     *
+     * @param hazelcastInstance The Hazelcast engine used by the application.
+     */
+    public HazelcastLockProvider(final HazelcastInstance hazelcastInstance) {
+        this(hazelcastInstance, LOCK_STORE_KEY_DEFAULT);
+    }
+
+    /**
+     * Instanciate the provider.
+     *
+     * @param hazelcastInstance The Hazelcast engine used by the application
+     * @param lockStoreKey      The key where the locks store is associate {@link #hazelcastInstance} (by default {@link #LOCK_STORE_KEY_DEFAULT}).
+     */
+    public HazelcastLockProvider(final HazelcastInstance hazelcastInstance, final String lockStoreKey) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.lockStoreKey = lockStoreKey;
+    }
+
+    @Override
+    public Optional<SimpleLock> lock(final LockConfiguration lockConfiguration) {
+        log.trace("lock - Attempt : {}", lockConfiguration);
+        final Instant now = Instant.now();
+        final String lockName = lockConfiguration.getName();
+        final IMap<String, HazelcastLock> store = getStore();
+        try {
+            // lock the map key entry
+            store.lock(lockName);
+            // just one thread at a time, in the cluster, can run this code
+            // each thread waits until the lock to be unlock
+            if (tryLock(lockConfiguration, now)) {
+                return Optional.of(() -> unlock(lockName));
```

Comment:
(Minor) Using lambda in this context is a bit confusing. It took me some time to figure out what it means. Would have been more readable to instantiate an implementation of SimpleLock which would contain the unlock method.